### PR TITLE
changed hkl specification for indexing to master hklID

### LINKED ISSUE #452: Change hkl indexing behavior in seeded search (far-field HEDM)
We used to live dangerously and have the `active_hkls` specified the indexing (`find-oreintations`) CLI workflow interpreted as direct indices into the `planeData.hkls` as generated.  This was bad for a few reasons:
1. The material is generated on load subject to minimum d-spacing and structure factor cutoff parameters, which can lead to non-sequential exclusions.
2. Non-sequential exclusions can also be saved in the materials archive, and will be merged on load with exclusions from 1.
Since we are selecting from the masked `planeData.hkls` blindly in the CLI (via the yaml config), this can lead to weird inconsistencies.

To remedy this, I propose using the unique `hklID` assigned to each hkl in the master `planeData.hklDataList` instead.  Ideally the config can be modified to take either this or an explicit list of hkls. 

For example, the single GE ruby config will look like:
```
find_orientations:
  orientation_maps:
    # A file name must be specified. If it doesn't exist, one will be created
    file: null

    threshold: 25 # threshold on raw images to build maps
    bin_frames: 1 # how many successive frames to bin. default is 1

    # specify which hkls to use
    # "all", or a list of hkl indices.  Note that the default value of "all"
    # means all hkls that are not marked as excluded in the material definition.
    # can be a list of indices or a list of [h, k, l].
    # active_hkls: [0,1,2,4,6,7]
    active_hkls:
      - [0, 1, 2]
      - [1, 0, 4]
      - [1, 1, 0]
      - [1, 1, 3]
      - [0, 2, 4]
      - [1, 1, 6]
```

@bnmajor , @psavery , and @cjh1 -- This` will finally allow us to clean up https://github.com/HEXRD/hexrdgui/pull/970; we can force it to write out the explicit hkls used as active hkls.

### PR DESCRIPTION
**Pull request by joelvbernier**
This closes #452 and cleans up the yaml config for far-field HEDM.

- [x] Change `PlaneData.getHKL` to take an optional `hklID` args
- [x] Change `PlaneData.getHKLID` to optionally return reduced or master indices
- [x] Change `EtaOmeMaps.iHKLList` to reflect master `hklID`
- [x] Track down all uses of `EtaOmeMaps.iHKLList` and reconcile the update to master indices.

We will probably have to make accompanying changes in `hexrdgui` as well @psavery...

Fixes: #188

### CODE REVIEW

**Comment by joelvbernier** [comment]
Ah -- the eta-ome map comparison test will fail until we update the [config in the repo](https://github.com/HEXRD/examples/pull/2) to reflect the changes to `active_hkl` specification.

**Comment by joelvbernier** [comment]
This PR also closes #188

File: hexrd/findorientations.py [hunk 1]
```diff
@@ -423,12 +425,11 @@ def load_eta_ome_maps(cfg, pd, image_series, hkls=None, clean=False):
...
             res = EtaOmeMaps(fn)
             pd = res.planeData
-            available_hkls = pd.hkls.T
             logger.info('loaded eta/ome orientation maps from %s', fn)
-            hkls = [str(i) for i in available_hkls[res.iHKLList]]
+            shkls = pd.getHKLs(*res.iHKLList, asStr=True)
             logger.info(
                 'hkls used to generate orientation maps: %s',
-                hkls
+                [f'[{i}]' for i in shkls]
```
Comment (psavery) [inline hexrd/findorientations.py:432]: Nice!

**Comment by psavery** [comment]
I know we'll need to update hexrdgui [here](https://github.com/HEXRD/hexrdgui/blob/a3e45bc324d14560cf574cde5bdb8f28dffd803d/hexrd/ui/indexing/create_config.py#L28), and possibly at other places as well.

I'll give this a try in hexrdgui soon and put up a corresponding hexrdgui PR.

**Comment by joelvbernier** [comment]
I just realized that the `EtaOmeMaps.save_eta_ome_maps` method was only exporting the reduced list of hkls, so that on load it would have a completely different master list (literally only containing the `active_hkls` from the map generation).  This creates an obvious inconsistency!  I now save the full list of hkls on the input `PlaneData` using the `hklDataList` attribute.  I _*think*_ this fixes the problem with trying to index off of a previously saved map instead of generating a new one.  I just tested this.

File: hexrd/config/findorientations.py [hunk 2]
```diff
@@ -217,6 +217,12 @@ def active_hkls(self):
             temp = None
         return temp
 
+    @property
+    def explicit_hkls(self):
+        return self._cfg.get(
+            'find_orientations:orientation_maps:explicit_hkls', default=1
+            )
```
[the remark this answers is not in the document]
Comment (joelvbernier) [reply hexrd/config/findorientations.py:224]: Ah -- I need to remove this.  I initially thought I was going to support the old was of using live (reduced) indices, but I decided against it.  I'll push the changes.

File: hexrd/findorientations.py [hunk 3]
```diff
@@ -84,20 +84,27 @@ def generate_orientation_fibers(cfg, eta_ome):
...
     tTh = pd.getTTh()
     bMat = pd.latVecOps['B']
     csym = pd.getLaueGroup()
 
+    # !!! changed recently where iHKLList are now master hklIDs
+    pd_hkl_ids = eta_ome.iHKLList[seed_hkl_ids]
+    pd_hkl_idx = pd.getHKLID(
+        pd.getHKLs(*eta_ome.iHKLList).T,
+        master=False
+    )
```
[the remark this answers is not in the document]
Comment (joelvbernier) [reply hexrd/findorientations.py:98]: Probably ;-)

File: hexrd/config/findorientations.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (joelvbernier) [reply hexrd/config/findorientations.py]: fixed now in d27527271b1e9771058a66959d51811422cd8dd2

**Comment by joelvbernier** [comment]
Now the `PlaneData` object on the re-instantiated (from saved) eta-ome maps should be functionally identical to the one used in the generation.  Exclusions were previously missing.

**Review by psavery — APPROVED** [review]
@joelvbernier Okay, the tests appear to be working now with the changes I added to hexrd/examples#2. However, perhaps we can wait to merge until we have a corresponding hexrdgui PR up as well?

**Review by psavery — APPROVED** [review]
After we merge hexrd/examples#2, I think this is good to go.

**Comment by psavery** [comment]
@joelvbernier I re-ran the CI after we merged the examples repo updates, and it looks like it passes now!

**Review by psavery — APPROVED** [review]
LGTM

**Pull request merged**